File: workers/worker-c-cpp/src/worker_runner.py
import os
import redis
from openregex_libs.lifecycle import start_heartbeat, install_shutdown_unregister
from worker_c_cpp.engines import WORKER_INFO
from worker_c_cpp.registry import register_engines
from worker_c_cpp.processor import listen_and_process


def main():
    redis_url = os.environ.get("REDIS_URL", "redis://redis:6379")
    redis_client = redis.Redis.from_url(redis_url, decode_responses=True)

    # Heartbeat must be live before registration so discovery never sees
    # a registered worker without one.
    start_heartbeat(redis_client, WORKER_INFO.worker_name)
    register_engines(redis_client)
    install_shutdown_unregister(redis_client, WORKER_INFO.worker_name)
    listen_and_process(redis_client)


if __name__ == "__main__":
    main()

File: workers/worker-v8/src/worker_runner.js
import { Redis } from "ioredis";
import { registerEngines } from "./registry.js";
import { listenAndProcess } from "./processor.js";
import { WORKER_INFO } from "./engines/index.js";

const WORKERS_HASH_KEY = "openregex:workers";
const HEARTBEAT_KEY = `openregex:workers:heartbeat:${WORKER_INFO.worker_name}`;
const HEARTBEAT_TTL_S = 15;
const HEARTBEAT_INTERVAL_MS = 5000;

async function setHeartbeat(redisClient) {
  try {
    await redisClient.set(HEARTBEAT_KEY, Math.floor(Date.now() / 1000), "EX", HEARTBEAT_TTL_S);
  } catch {
    // transient Redis outage; the TTL just expires until it recovers
  }
}

function installShutdownUnregister(redisClient) {
  const shutdown = async () => {
    try {
      await redisClient.hdel(WORKERS_HASH_KEY, WORKER_INFO.worker_name);
      await redisClient.del(HEARTBEAT_KEY);
    } catch {
      // best effort
    }
    process.exit(0);
  };
  process.on("SIGTERM", shutdown);
  process.on("SIGINT", shutdown);
}

async function main() {
  const redisUrl = process.env.REDIS_URL || "redis://redis:6379";
  const redisClient = new Redis(redisUrl);
  const pubClient = new Redis(redisUrl);

  // Heartbeat must be live before registration so discovery never sees
  // a registered worker without one.
  await setHeartbeat(redisClient);
  await registerEngines(redisClient);
  setInterval(() => setHeartbeat(redisClient), HEARTBEAT_INTERVAL_MS);
  installShutdownUnregister(redisClient);

  await listenAndProcess(redisClient, pubClient);
}

main().catch(console.error);
